dali/kernels/imgproc/sampler.h

```diff
@@ -30,165 +30,183 @@ namespace kernels {
+  for (int i = 0; i < n; i++) {
+    if (static_cast<unsigned>(coords[i]) >= static_cast<unsigned>(limit[i]))
+      return false;
+  }
+  return true;
+}
+
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec2 coords, ivec2 limit) {
+  return static_cast<unsigned>(coords.x) < static_cast<unsigned>(limit.x) &&
+         static_cast<unsigned>(coords.y) < static_cast<unsigned>(limit.y);
+}
+
+/**
+ * @brief Checks if all coordinates in `coords` are positive and below the respective
+ *        coordinate in `limit`.
+ *
+ * @remarks `limit` is expected to have non-negative coordinates - otherwise, the result
+ *          is undefined.
+ */
+DALI_HOST_DEV DALI_FORCEINLINE
+constexpr bool all_in_range(ivec3 coords, ivec3 limit) {
+  return static_cast<unsigned>(coords.x) < static_cast<unsigned>(limit.x) &&
+         static_cast<unsigned>(coords.y) < static_cast<unsigned>(limit.y) &&
+         static_cast<unsigned>(coords.z) < static_cast<unsigned>(limit.z);
+}
+
+template <int _spatial_ndim, typename In>
+struct Sampler<DALI_INTERP_NN, _spatial_ndim, In> {
+  static constexpr int spatial_ndim = _spatial_ndim;
+
+  using icoords = dali::ivec<spatial_ndim>;
+  using fcoords = dali::vec<spatial_ndim>;
+
   Sampler() = default;
-  DALI_HOST_DEV explicit Sampler(const Surface2D<const In> &surface)
+  DALI_HOST_DEV explicit Sampler(const Surface<spatial_ndim, const In> &surface)
   : surface(surface) {}
 
-  Surface2D<const In> surface;
+  Surface<spatial_ndim, const In> surface;
 
   template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(
-      int x, int y, int c,
-      BorderValue border_value) const {
-    if (x < 0 || x >= surface.size.x ||
-        y < 0 || y >= surface.size.y) {
-      return ConvertSat<T>(GetBorderChannel(border_value, c));
+  DALI_HOST_DEV DALI_FORCEINLINE T at(icoords pos, int c, BorderValue border_value) const {
+    if (all_in_range(pos, surface.size)) {
+      return ConvertSat<T>(surface(pos, c));
     } else {
-      return ConvertSat<T>(surface(x, y, c));
+      return ConvertSat<T>(GetBorderChannel(border_value, c));
     }
   }
 
   template <typename T = In>
-  DALI_HOST_DEV T at(
-      int x, int y, int c,
-      BorderClamp) const {
-    ivec2 clamped = clamp(ivec2(x, y), ivec2(0, 0), surface.size - 1);
-    return ConvertSat<T>(surface(clamped, c));
-  }
-
-  template <typename T = In, typename BorderValue>
-  DALI_HOST_DEV T at(
-      float x, float y, int c,
-      BorderValue border_value) const {
-    return at<T>(floor_int(x), floor_int(y), c, border_value);
+  DALI_HOST_DEV DALI_FORCEINLINE T at(icoords pos, int c, BorderClamp) const {
+    return ConvertSat<T>(surface(clamp(pos, icoords(0), surface.size - 1), c));
```

Comment:
Why `clamped` intermediate value has been removed?

Reply:
I didn't give it much thought - the function was rewritten and the line didn't seem excessively long or complex with the call inlined.